From 03bc18e76289a74966292b30df5002f95695baf3 Mon Sep 17 00:00:00 2001
From: Alexandros Frantzis <alexandros.frantzis@collabora.com>
Date: Mon, 7 Oct 2024 18:38:46 +0300
Subject: [PATCH 13/14] media/gpu/v4l2: Use driver preferred format for encoder

Using other formats (even if in theory supported by the device)
seem to make the driver unhappy and fail when we try to enqueue
any buffers.
---
 media/gpu/v4l2/v4l2_video_encode_accelerator.cc | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/media/gpu/v4l2/v4l2_video_encode_accelerator.cc b/media/gpu/v4l2/v4l2_video_encode_accelerator.cc
index 3603bcaf93..bedae792ce 100644
--- a/media/gpu/v4l2/v4l2_video_encode_accelerator.cc
+++ b/media/gpu/v4l2/v4l2_video_encode_accelerator.cc
@@ -1831,13 +1831,13 @@ V4L2VideoEncodeAccelerator::NegotiateInputFormat(VideoPixelFormat input_format,
                << VideoPixelFormatToString(input_format);
     return std::nullopt;
   }
-  pix_fmt_candidates.push_back(input_fourcc->ToV4L2PixFmt());
-  // Second try preferred input formats for both single-planar and
-  // multi-planar.
+  // Use preferred input formats for both single-planar and multi-planar,
+  // to work around driver not working with requested input format.
   for (auto preferred_format :
        device_->PreferredInputFormat(V4L2Device::Type::kEncoder)) {
     pix_fmt_candidates.push_back(preferred_format);
   }
+  pix_fmt_candidates.push_back(input_fourcc->ToV4L2PixFmt());
 
   for (const auto pix_fmt : pix_fmt_candidates) {
     DVLOGF(3) << "Trying S_FMT with " << FourccToString(pix_fmt);
-- 
2.45.2

